diff -ruN linux-2.6.30.orig/drivers/net/phy/davicom.c linux-2.6.30/drivers/net/phy/davicom.c
--- linux-2.6.30.orig/drivers/net/phy/davicom.c	2009-06-09 21:05:27.000000000 -0600
+++ linux-2.6.30/drivers/net/phy/davicom.c	2010-07-19 13:02:06.824704289 -0600
@@ -172,6 +172,17 @@
 	.driver		= { .owner = THIS_MODULE,},
 };
 
+static struct phy_driver dm9161b_driver = {
+	.phy_id		= 0x0181b8b0,
+	.name		= "Davicom DM9161B",
+	.phy_id_mask	= 0x0ffffff0,
+	.features	= PHY_BASIC_FEATURES,
+	.config_init	= dm9161_config_init,
+	.config_aneg	= dm9161_config_aneg,
+	.read_status	= genphy_read_status,
+	.driver		= { .owner = THIS_MODULE,},
+};
+
 static struct phy_driver dm9131_driver = {
 	.phy_id		= 0x00181b80,
 	.name		= "Davicom DM9131",
@@ -197,11 +208,17 @@
 	if (ret)
 		goto err2;
 
-	ret = phy_driver_register(&dm9131_driver);
+	ret = phy_driver_register(&dm9161b_driver);
 	if (ret)
 		goto err3;
+
+	ret = phy_driver_register(&dm9131_driver);
+	if (ret)
+		goto err4;
 	return 0;
 
+ err4:
+	phy_driver_unregister(&dm9161b_driver);
  err3:
 	phy_driver_unregister(&dm9161a_driver);
  err2:
@@ -213,6 +232,7 @@
 static void __exit davicom_exit(void)
 {
 	phy_driver_unregister(&dm9161e_driver);
+	phy_driver_unregister(&dm9161b_driver);
 	phy_driver_unregister(&dm9161a_driver);
 	phy_driver_unregister(&dm9131_driver);
 }
