diff -ruN at91bootstrap-2.11.orig/board/at91sam9g20ek/at91sam9g20ek.c at91bootstrap-2.11/board/at91sam9g20ek/at91sam9g20ek.c
--- at91bootstrap-2.11.orig/board/at91sam9g20ek/at91sam9g20ek.c	2010-10-27 13:12:15.973699072 -0600
+++ at91bootstrap-2.11/board/at91sam9g20ek/at91sam9g20ek.c	2010-10-27 13:31:53.760666136 -0600
@@ -67,6 +67,8 @@
 		{"RXD", AT91C_PIN_PB(14), 0, PIO_DEFAULT, PIO_PERIPH_A},
 		{"TXD", AT91C_PIN_PB(15), 0, PIO_DEFAULT, PIO_PERIPH_A},
 #endif
+		{"USBEN",  AT91C_PIN_PA(22), 0, PIO_DEFAULT, PIO_INPUT},
+		{"PLED",  AT91C_PIN_PC(15), 0, PIO_DEFAULT, PIO_OUTPUT},
 		{(char *) 0, 0, 0, PIO_DEFAULT, PIO_PERIPH_A},
 	};
 
@@ -227,20 +229,20 @@
 /*------------------------------------------------------------------------------*/
 static void nand_recovery(void)
 {
-	/* Configure PIOs */
-	const struct pio_desc bp4_pio[] = {
-		{"BP4", AT91C_PIN_PA(31), 0, PIO_PULLUP, PIO_INPUT},
-		{(char *) 0, 0, 0, PIO_DEFAULT, PIO_PERIPH_A},
-	};
-
-	/* Configure the PIO controller */
-	writel((1 << AT91C_ID_PIOA), PMC_PCER + AT91C_BASE_PMC);
-	pio_setup(bp4_pio);
-	
-	/* If BP4 is pressed during Boot sequence */
-	/* Erase NandFlash block 0*/
-	if (!pio_get_value(AT91C_PIN_PA(31)) )
-		AT91F_NandEraseBlock0();
+//	/* Configure PIOs */
+//	const struct pio_desc bp4_pio[] = {
+//		{"BP4", AT91C_PIN_PA(31), 0, PIO_PULLUP, PIO_INPUT},
+//		{(char *) 0, 0, 0, PIO_DEFAULT, PIO_PERIPH_A},
+//	};
+//
+//	/* Configure the PIO controller */
+//	writel((1 << AT91C_ID_PIOA), PMC_PCER + AT91C_BASE_PMC);
+//	pio_setup(bp4_pio);
+//	
+//	/* If BP4 is pressed during Boot sequence */
+//	/* Erase NandFlash block 0*/
+//	if (!pio_get_value(AT91C_PIN_PA(31)) )
+//		AT91F_NandEraseBlock0();
 }
 
 /*------------------------------------------------------------------------------*/
